### PR TITLE
*: add builtin function UNIX_TIMESTAMP

### PR DESCRIPTION
**Pull request by tiancaiamao**
For issue https://github.com/pingcap/tidb/issues/236

### CODE REVIEW

File: expression/builtin_time.go [hunk 1]
```diff
@@ -745,3 +745,43 @@ func parseDayInterval(sc *variable.StatementContext, value types.Datum) (int64,
 	}
 	return value.ToInt64(sc)
 }
+
+func builtinUnixTimestamp(args []types.Datum, ctx context.Context) (d types.Datum, err error) {
```
Comment (shenli) [inline expression/builtin_time.go]: Add reference.

File: expression/builtin_time.go [hunk 2]
```diff
@@ -745,3 +745,43 @@ func parseDayInterval(sc *variable.StatementContext, value types.Datum) (int64,
...
+	switch args[0].Kind() {
+	case types.KindString:
+		t, e = types.ParseTime(args[0].GetString(), mysql.TypeDatetime, types.MaxFsp)
+	case types.KindInt64, types.KindUint64:
+		t, e = types.ParseTimeFromInt64(args[0].GetInt64())
+	default:
+		return
+	}
+	if e != nil {
+		return
```
Comment (shenli) [inline expression/builtin_time.go]: return d, errors.Trace(e)?

File: parser/parser.y [hunk 3]
```diff
@@ -278,6 +278,7 @@ import (
 	trim		"TRIM"
 	rtrim 		"RTRIM"
 	ucase 		"UCASE"
+	unixTimestamp	"UNIX_TIMESTAMP"
```
Comment (shenli) [inline parser/parser.y]: Is UNIX_TIMESTAMP a keyword or non-keyword-token?

**Comment by shenli** [comment]
@tiancaiamao Please fix CI.

File: parser/parser.y (hunk 3 above)
Comment (tiancaiamao) [inline parser/parser.y]: it's not a keyword or non-keyword.
https://dev.mysql.com/doc/refman/5.7/en/keywords.html

File: expression/builtin_time.go (hunk 2 above)
Comment (tiancaiamao) [inline expression/builtin_time.go]: I find mysql default behavior is return 0, when input data is wrong...

**Comment by tiancaiamao** [comment]
PTAL

**Comment by zimulala** [comment]
@tiancaiamao  Please check if we should update type inferred? 
And fix the conflict.

**Comment by zimulala** [comment]
LGTM

File: expression/builtin_time.go [hunk 4]
```diff
@@ -746,3 +751,52 @@ func parseDayInterval(sc *variable.StatementContext, value types.Datum) (int64,
...
+		now := time.Now().Unix()
+		d.SetInt64(now)
+		return
+	}
+
+	d.SetInt64(0)
+	var (
+		t  types.Time
+		t1 time.Time
+		e  error
```
Comment (hanfei1991) [inline expression/builtin_time.go]: There exists an err so you needn't to declare an error e.

**Comment by shenli** [comment]
Ping @tiancaiamao

File: expression/builtin_time.go (hunk 4 above)
Comment (tiancaiamao) [inline expression/builtin_time.go]: I know, but if I use `if err != nil { return }` , this function would return error, which is not the same as mysql behavior.

**Comment by hanfei1991** [comment]
LGTM

**Comment by tiancaiamao** [comment]
For invalid input, `select UNIX_TIMESTAMP(xxx)`  will return 0 and have warning, insert that value to column would get error.
I change the code to return error when mysql have warning.
PTAL @shenli

**Pull request merged**